src/Storages/MergeTree/MergeTreeIndexSimpleHnsw.cpp

```diff
@@ -0,0 +1,172 @@
+#include <cstddef>
+#include <cstring>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <spacefactory.h>
+#include <Storages/MergeTree/MergeTreeIndexSimpleHnsw.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+MergeTreeIndexGranuleSimpleHnsw::MergeTreeIndexGranuleSimpleHnsw(const String & index_name_, 
+    const Block & index_sample_block_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+{}
+
+
+MergeTreeIndexGranuleSimpleHnsw::MergeTreeIndexGranuleSimpleHnsw(const String & index_name_,
+    const Block & index_sample_block_, 
+    std::unique_ptr<similarity::Hnsw<float>> index_impl_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_), index_impl(std::move(index_impl_))
+{}
+
+
+void MergeTreeIndexGranuleSimpleHnsw::serializeBinary(WriteBuffer & ostr) const {
+     std::ostringstream ost;
+     index_impl->SaveIndex(ost);
+     ostr.write(ost.str().data(), ost.str().size());
+}
+
+void MergeTreeIndexGranuleSimpleHnsw::deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion version){
+    if (version != 1){
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown index version {}.", version);
+    }
+    size_t index_size = istr.available();
+    std::string str(index_size, '\0');
+    istr.read(str.data(), index_size);
+    std::istringstream ist(str);
+    index_impl->LoadIndex(ist);
+}
+
+MergeTreeIndexAggregatorSimpleHnsw::MergeTreeIndexAggregatorSimpleHnsw(const String & index_name_, 
+const Block & index_sample_block_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+{}
+
+MergeTreeIndexGranulePtr MergeTreeIndexAggregatorSimpleHnsw::getGranuleAndReset()
+{
+    std::string space_type = "l2";
+    auto space_params = std::shared_ptr<similarity::AnyParams>(new similarity::AnyParams(std::vector<std::string>()));
+    auto* space(similarity::SpaceFactoryRegistry<float>::Instance().CreateSpace(space_type, *space_params));
+    auto index_impl = std::make_unique<similarity::Hnsw<float>>(false, *space, data);
+    return std::make_shared<MergeTreeIndexGranuleSimpleHnsw>(index_name, index_sample_block, std::move(index_impl));
+}
+
+void MergeTreeIndexAggregatorSimpleHnsw::update(const Block & block, size_t * pos, size_t limit){
+     if (*pos >= block.rows())
+        throw Exception(
+                "The provided position is not less than the number of block rows. Position: "
+                + toString(*pos) + ", Block rows: " + toString(block.rows()) + ".", ErrorCodes::LOGICAL_ERROR);
+
+    size_t rows_read = std::min(limit, block.rows() - *pos);
+
+   auto index_column_name = index_sample_block.getByPosition(0).name;
+    const auto & column = block.getByName(index_column_name).column->cut(*pos, rows_read);
+
+    for (size_t i = 0; i < rows_read; ++i) {
```

Comment:
Same as https://github.com/Vector-Similarity-Search-for-ClickHouse/ClickHouse/pull/24#discussion_r837305594